File: LLM-Translations/GPT-4/codenet/C++/Go/s143031369.go

package main

import (
	"fmt"
	"math"
	"sort"
)

const eps = 1e-8

type Point struct {
	x, y float64
}

func sig(d float64) int {
	if math.Abs(d) < eps {
		return 0
	}
	if d < 0 {
		return -1
	}
	return 1
}

func dot(o, a, b Point) float64 {
	return (a.x-o.x)*(b.x-o.x) + (a.y-o.y)*(b.y-o.y)
}

func cross(o, a, b Point) float64 {
	return (a.x-o.x)*(b.y-o.y) - (b.x-o.x)*(a.y-o.y)
}

func btw(x, a, b Point) int {
	return sig(dot(x, a, b))
}

func dis(a, b Point) float64 {
	return math.Sqrt((a.x-b.x)*(a.x-b.x) + (a.y-b.y)*(a.y-b.y))
}

func cos(o, a, b Point) float64 {
	return dot(o, a, b) / dis(o, a) / dis(o, b)
}

func jarvis(p []Point, n int) []int {
	ch := make([]int, n)
	minIdx := 0
	for i := 1; i < n; i++ {
		if p[i].x < p[minIdx].x {
			minIdx = i
		}
	}
	ch[0] = minIdx
	l := minIdx
	d := 1
	for {
		o := l
		for i := 0; i < n; i++ {
			t := sig(cross(p[o], p[l], p[i]))
			if t > 0 || (t == 0 && btw(p[l], p[o], p[i]) <= 0) {
				l = i
			}
		}
		ch[d] = l
		d++
		if l == ch[0] {
			break
		}
	}
	return ch[:d-1]
}

func main() {
	var n, xx, yy int
	fmt.Scan(&n)
	p := make([]Point, n)
	ang := make([]float64, n)
	ans := make([]float64, n)

	for i := 0; i < n; i++ {
		fmt.Scan(&xx, &yy)
		p[i] = Point{float64(xx), float64(yy)}
	}

	if n == 2 {
		fmt.Println("0.5")
		fmt.Println("0.5")
	} else {
		ch := jarvis(p, n)
		if len(ch) == 2 {
			ans[ch[0]] = 0.5
			ans[ch[1]] = 0.5
		} else {
			for _, idx := range ch {
				i := sort.SearchInts(ch, (idx-1+len(ch))%len(ch))
				j := sort.SearchInts(ch, (idx+1)%len(ch))
				ang[idx] = math.Pi - math.Acos(cos(p[idx], p[ch[i]], p[ch[j]]))
			}
			for i := 0; i < n; i++ {
				ans[i] = ang[i] / 2.0 / math.Pi
			}
		}
		for i := 0; i < n; i++ {
			fmt.Printf("%.20f\n", ans[i])
		}
	}
}

File: LLM-Translations/GPT-4/codenet/C++/Python/s143031369.py

import math

def sig(d):
    return 0 if abs(d) < 1e-8 else -1 if d < 0 else 1

class Point:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

    def set(self, x, y):
        self.x = x
        self.y = y

    def mod(self):
        return math.sqrt(self.x * self.x + self.y * self.y)

    def mod_pow(self):
        return self.x * self.x + self.y * self.y

    def __lt__(self, other):
        return sig(self.x - other.x) != 0 if self.x < other.x else sig(self.y - other.y) < 0


def dot(o, a, b):
    return (a.x - o.x) * (b.x - o.x) + (a.y - o.y) * (b.y - o.y)


def cross(o, a, b):
    return (a.x - o.x) * (b.y - o.y) - (b.x - o.x) * (a.y - o.y)


def dis(a, b):
    return math.sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y))


def cos(o, a, b):
    return dot(o, a, b) / dis(o, a) / dis(o, b)


def jarvis(p, n, ch):
    d, i = 0, 0
    for i in range(n):
        if p[i] < p[d]:
            d = i
    l = s = ch[0] = d
    d = 1
    while True:
        o = l
        for i in range(n):
            t = sig(cross(p[o], p[l], p[i]))
            if t > 0 or (t == 0 and dot(p[l], p[o], p[i]) <= 0):
                l = i
        ch[d] = l
        d += 1
        if l == s:
            break
    return d - 1


def main():
    n = int(input())
    P = [Point() for _ in range(n)]
    for i in range(n):
        x, y = map(int, input().split())
        P[i].set(x, y)
    
    ang = [0] * n
    ans = [0] * n
    ch = [0] * n

    if n == 2:
        print("0.5")
        print("0.5")
    else:
        m = jarvis(P, n, ch)
        if m == 2:
            ans[ch[0]] = 0.5
            ans[ch[1]] = 0.5
        else:
            for i in range(m):
                ang[ch[i]] = math.pi - math.acos(cos(P[ch[i]], P[ch[(i - 1 + m) % m]], P[ch[(i + 1) % m]]))
            
            for i in range(n):
                ans[i] = ang[i] / 2.0 / math.pi

        for i in range(n):
            print("{:.20f}".format(ans[i]))

if __name__ == "__main__":
    main()



